### PR TITLE
[WebPubSub] Adding feature support

### PR DESCRIPTION
**Pull request by vicancy**
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

Supporting new features in Web PubSub
1. Add "groups" parameter when generating client access URI
2. Add a helper class `SearchFilter` for generating the filter string, referencing the practice of Cognitive Search
3. Add more tests

### CODE REVIEW

File: sdk/webpubsub/Azure.Messaging.WebPubSub/api/Azure.Messaging.WebPubSub.netstandard2.0.cs [hunk 1]
```diff
@@ -1,5 +1,10 @@
 namespace Azure.Messaging.WebPubSub
 {
+    public static partial class SearchFilter
```
Comment (ArthurMa1978) [inline sdk/webpubsub/Azure.Messaging.WebPubSub/api/Azure.Messaging.WebPubSub.netstandard2.0.cs:3]: The name is too generic, consider adding a prefix

File: sdk/webpubsub/Azure.Messaging.WebPubSub/api/Azure.Messaging.WebPubSub.netstandard2.0.cs [hunk 2]
```diff
@@ -33,10 +38,14 @@ public partial class WebPubSubServiceClient
...
-        public virtual System.Threading.Tasks.Task<System.Uri> GetClientAccessUriAsync(System.DateTimeOffset expiresAt, string userId = null, System.Collections.Generic.IEnumerable<string> roles = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
-        public virtual System.Threading.Tasks.Task<System.Uri> GetClientAccessUriAsync(System.TimeSpan expiresAfter = default(System.TimeSpan), string userId = null, System.Collections.Generic.IEnumerable<string> roles = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        public virtual System.Uri GetClientAccessUri(System.DateTimeOffset expiresAt, string userId = null, System.Collections.Generic.IEnumerable<string> roles = null, System.Collections.Generic.IEnumerable<string> groups = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
```
Comment (ArthurMa1978) [inline sdk/webpubsub/Azure.Messaging.WebPubSub/api/Azure.Messaging.WebPubSub.netstandard2.0.cs:41]: Here introduced breaking changes which is not allowed if not increase the major version.
2 solutions:
1. Increase the version to 2.0.0
2. Keep old methods and add new overloads

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/SearchFilter.cs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        /// </summary>
+        /// <param name="text">The text to quote.</param>
+        /// <returns>The quoted text.</returns>
+        private static string Quote(string text)
+        {
+            if (text == null)
+            { return "null"; }
+
+            // Optimistically allocate an extra 5% for escapes
+            StringBuilder builder = new StringBuilder(2 + (int)(text.Length * 1.05));
```
Comment (ArthurMa1978) [inline sdk/webpubsub/Azure.Messaging.WebPubSub/src/SearchFilter.cs:99]: How about use the [String.Replace](https://learn.microsoft.com/en-us/dotnet/api/system.string.replace?view=net-7.0) directly?

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_helpers.cs [hunk 4]
```diff
@@ -6,6 +6,7 @@
 using System.Linq;
 using System.Text;
 using System.Text.Json;
+using System.Text.RegularExpressions;
```
Comment (ArthurMa1978) [inline sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_helpers.cs:9]: I didn't find any regex functionality be used in this file

File: sdk/webpubsub/Azure.Messaging.WebPubSub/api/Azure.Messaging.WebPubSub.netstandard2.0.cs (hunk 2 above)
Comment (vicancy) [reply sdk/webpubsub/Azure.Messaging.WebPubSub/api/Azure.Messaging.WebPubSub.netstandard2.0.cs:41]: I don't think it is a breaking change though, the previous method (L36) is changed to L42 with all the default values removed. I thought it is the recommended practice to add the parameter `groups` to the method having default values. Please correct me if I am wrong.

File: sdk/webpubsub/Azure.Messaging.WebPubSub/api/Azure.Messaging.WebPubSub.netstandard2.0.cs (hunk 1 above)
Comment (vicancy) [reply sdk/webpubsub/Azure.Messaging.WebPubSub/api/Azure.Messaging.WebPubSub.netstandard2.0.cs:3]: Referenced https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/search/Azure.Search.Documents/src/SearchFilter.cs#L19

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/SearchFilter.cs (hunk 3 above)
Comment (vicancy) [reply sdk/webpubsub/Azure.Messaging.WebPubSub/src/SearchFilter.cs:99]: Referenced https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/search/Azure.Search.Documents/src/SearchFilter.cs#L19

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_helpers.cs (hunk 4 above)
Comment (vicancy) [reply sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_helpers.cs:9]: nice catch

File: sdk/webpubsub/Azure.Messaging.WebPubSub/api/Azure.Messaging.WebPubSub.netstandard2.0.cs (hunk 2 above)
Comment (ArthurMa1978) [reply sdk/webpubsub/Azure.Messaging.WebPubSub/api/Azure.Messaging.WebPubSub.netstandard2.0.cs:41]: Got confirmed from @m-nash, this is not breaking changes, case close :)

**Review by ArthurMa1978 — APPROVED** [review]

**Comment by vicancy** [comment]
Hi @jsquire do you have any expert looking into the CI failure on "Build Test Windows2019_NET461_ProjectRef_Release" complaining about some assembly missing? This is strange and should not be related to my change because my previous commit passed successfully however the next commit containing only a simple rename failed with this CI failure.

**Comment by jsquire** [comment]
@vicancy:  I don't see any failures.   I'd suggest raising issues around pipelines in the "Engineering System" channel on Teams so that the right folks can assist.

**Comment by vicancy** [comment]
Thanks for the review @ArthurMa1978 ! Could you also help approve the APIview https://apiview.dev/Assemblies/Review/c02c4162593a42a7b404ca7d927e4cef?

**Pull request merged**